packages/react-aria-components/src/utils.tsx

```diff
@@ -225,6 +225,10 @@ export function useExitAnimation(ref: RefObject<HTMLElement>, isOpen: boolean) {
 function useAnimation(ref: RefObject<HTMLElement>, isActive: boolean, onEnd: () => void) {
   let prevAnimation = useRef<string | null>(null);
   if (isActive && ref.current) {
+    // This is ok because we only read it in the layout effect below, immediately after the commit phase.
```

Comment:
> i think the bigger issue with this is that we shouldn't be reading from the dom during render

but why? We already guarantee it's in the DOM due to the `ref.current` check here.

> technically this is already running every render

no, it is conditional if `isActive && ref.current`, so it only runs when the animation starts. if we moved it to an effect, it would need to run every time, even while the animation is not active. Waiting until it becomes active would be too late because the animation style would already be updated at that point.

I agree it's a little weird, but I think this is the most efficient way of doing it. Ideally there'd be a type of effect that ran just _before_ the DOM was mutated, but there isn't. Render is the closest we have.

Reply:
Ah, I see. Thanks for explaining.